### PR TITLE
MAINT: special: remove more `npy_math.h` usage

### PR DESCRIPTION
**Pull request by rgommers**
Cephes is now free of all `npy_*` and `NPY_*` symbols. This reduces usage down to the not-yet-replaceable typedefs and complex dtype functions.

### CODE REVIEW

**Comment by rgommers** [comment]
This fixes an issue with the use of `npymath` that turned out to be a problem in numpy 1.24.0rc1 (use of `npy_asinh`, removed here). The symbol will be restored in numpy through a compat shim, but still we should get rid of it. I marked this for `1.10.0`. The goal for 1.11.0 should be to stop using `libnpymath.a` and `libnpyrandom.a` completely.

**Comment by WarrenWeckesser** [comment]
Sorry I haven't been following the `npymath` issues closely.  Is there an "executive summary" of what is going on?  Is the idea that everything that we were using from `npymath` is now available in the standard C/C++ libraries, so `npymath` is obsolete?

**Comment by rgommers** [comment]
> Is the idea that everything that we were using from `npymath` is now available in the standard C/C++ libraries, so `npymath` is obsolete?

Close to it - whatever is in C99 _and_ universally supported by compilers we care about is what we should use. The main exception is complex types, which aren't supported by MSVC, so we still need a little compat layer for that.


> Is there an "executive summary" of what is going on?

The main issue is using a static library at all. This is bad, mainly for these two reasons:
1. It is not guaranteed that `numpy` and `scipy` are built with the same compilers. And that leads to potential issues that are hard to predict, as we've experienced multiple times already. Especially on Windows, where we build scipy with mingw and numpy with MSVC. That was the reason why using the `vc142` instead of the `vc141` toolchain in numpy 1.21.5-1.22.1 led to complex issues and that change had to be reverted.
2. Cross-compiling tends to break. We haven't officially supported that in the past, but even in the one case where we do it (macOS x64-64 to arm64) it goes pretty horribly wrong: we stuff the x86-64 `libnpymath.a` into the macOS scipy wheel. This silently worked, because the required symbols were already loaded by numpy itself. Now that they're being removed from `libnpymath.a`, we're all of a sudden loading them from `libnpymath.a` and they have the wrong architecture.

tl;dr we should vendor the sources for the few remaining things we need, rather than use numpy's static libraries.

File: scipy/special/cephes/mconf.h [hunk 1]
```diff
@@ -100,7 +99,12 @@
...
 #define cephes_isnan(x) isnan(x)
 #define cephes_isinf(x) isinf(x)
-#define cephes_isfinite(x) npy_isfinite(x)
+#define cephes_isfinite(x) isfinite(x)
 #endif
 
+/* Constants needed that are not available in the C standard library */
+#define SCIPY_EULER     0.577215664901532860606512090082402431   /* Euler constant */
+#define SCIPY_El        2.718281828459045235360287471352662498L  /* e as long double */
+#define SCIPY_LOGE2     0.693147180559945309417232121458176568   /* log_e 2 */
```
Comment (WarrenWeckesser) [inline scipy/special/cephes/mconf.h:108]: With [Microsoft compilers](https://learn.microsoft.com/en-us/cpp/c-runtime-library/math-constants?view=msvc-170) and [GNU compilers](https://www.gnu.org/software/libc/manual/html_node/Mathematical-Constants.html), this is `M_LN2`.  Can we use that instead of defining `SCIPY_LOGE2`?

Comment (rgommers) [reply scipy/special/cephes/mconf.h:108]: Good catch. Yes, I think so - will update.

**Comment by rgommers** [comment]
There's a more complete summary in gh-17498 now.

**Review by tylerjereddy — APPROVED** [review]
Ok, I read the full diff and Warren's comment was addressed.

The CI is passing (1 failure already fixed in `main` with setuptools pin), and the changes make sense.
